src/videoframeprocessor/DefaultVideoTransformDevice.ts

```diff
@@ -0,0 +1,231 @@
+    }
+
+    // Device ID.
+    if (typeof this.device === 'string') {
+      if (this.browserBehavior.requiresNoExactMediaStreamConstraints()) {
+        trackConstraints.deviceId = this.device;
+      } else {
+        trackConstraints.deviceId = { exact: this.device };
+      }
+      return trackConstraints;
+    }
+
+    if ((this.device as MediaStream).id) {
+      // Nothing we can do.
+      return this.device;
+    }
+
+    // It's constraints.
+    return {
+      ...this.device,
+      ...trackConstraints,
+    };
+  }
+
+  /**
+   * Create {@link VideoFrameProcessorPipeline} if there is not a existing one and start video processors.
+   * Returns output `MediaStream` produced by {@link VideoFrameProcessorPipeline}.
+   */
+  async applyProcessors(mediaStream?: MediaStream): Promise<MediaStream> {
+    await this.createVideoFrameProcessorPipeline();
+    await this.pipe.setInputMediaStream(mediaStream);
+    this.inputMediaStream = mediaStream;
+    return this.pipe.outputMediaStream;
+  }
+
+  /**
+   * onOutputStreamDisconnect is called when device controller wants to detach
+   * the transform device. The default behavior is to stop the output
+   * media stream and release the input the media stream. If the input media stream
+   * is the provided device, it will not be released.
+   */
+  onOutputStreamDisconnect(): void {
+    this.logger.info('DefaultVideoTransformDevice: detach stopping input media stream');
+
+    const deviceIsMediaStream = this.device && (this.device as MediaStream).id;
+
+    // Stop processing but keep the pipe and processors
+    if (this.pipe) {
+      this.pipe.stop();
+    }
+
+    // Turn off the camera, unless device is a MediaStream
+    if (!deviceIsMediaStream) {
+      if (this.inputMediaStream) {
+        for (const track of this.inputMediaStream.getVideoTracks()) {
+          track.stop();
+        }
+      }
+    }
+  }
+
+  /**
+   * Dispose of the inner workings of the transform device, including pipeline and processors.
+   * `stop` can only be called when the transform device is not used by device controller anymore.
+   * After `stop` is called, all transform devices which share the pipeline must be discarded.
+   */
+  async stop(): Promise<void> {
+    if (this.inputMediaStream) {
+      for (const track of this.inputMediaStream.getVideoTracks()) {
+        track.stop();
+      }
+    }
+    if (this.pipe) {
+      await this.pipe.setInputMediaStream(null);
+    }
+
+    if (this.processors) {
+      for (const proc of this.processors) {
+        await proc.destroy();
+      }
+    }
+
+    this.inputMediaStream = null;
+    this.pipe = null;
+    this.processors = null;
+  }
+
+  private async createVideoFrameProcessorPipeline(): Promise<VideoFrameProcessorPipeline> {
+    if (!this.pipe) {
+      this.pipe = new DefaultVideoFrameProcessorPipeline(this.logger);
+      this.pipe.addObserver(this);
+    }
+
+    this.pipe.processors = this.processors;
+
+    return this.pipe;
+  }
```

Comment:
Why doesn't the pipeline own the processors?

Reply:
updated.  `pipe.destroy` will destroy the processors,  so `transformDevice.stop` will release pipe + processors, does this help ?